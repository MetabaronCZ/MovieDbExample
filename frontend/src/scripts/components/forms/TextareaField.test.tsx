import { act } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';
import { fireEvent, getByRole, getByText } from '@testing-library/dom';

import { TextareaField } from 'components/forms/TextareaField';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

describe('components/forms/TextareaField', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <TextareaField
        id="ID"
        label="LABEL"
        placeholder="PLACEHOLDER"
        name="NAME"
        value="VALUE"
        height={1}
        maxLength={100}
        required
        onChange={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const label = getByText(container, 'LABEL');
    expect(label).toBeInTheDocument();
    expect(label.tagName.toLowerCase()).toEqual('label');
    expect(label).toHaveAttribute('for', 'ID');

    const required = label.querySelector('span');
    expect(required).toBeInTheDocument();

    const textarea = getByRole(container, 'textbox');
    expect(textarea).toHaveAttribute('id', 'ID');
    expect(textarea).toHaveTextContent('VALUE');
    expect(textarea).toHaveAttribute('placeholder', 'PLACEHOLDER');
    expect(textarea).toHaveAttribute('name', 'NAME');
    expect(textarea).toHaveAttribute('maxlength', '100');
    expect(textarea).toHaveStyle({ height: 'calc(16px)' });
  });

  it('should render error', async () => {
    const { container } = await waitForComponent(
      <TextareaField
        label="LABEL"
        error="SOME-ERROR"
        value="VALUE"
        onChange={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const error = getByText(container, 'SOME-ERROR');
    expect(error).toBeInTheDocument();
  });

  it('should render info text', async () => {
    const { container } = await waitForComponent(
      <TextareaField
        label="LABEL"
        info="SOME-INFO"
        value="VALUE"
        onChange={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const info = getByText(container, 'SOME-INFO');
    expect(info).toBeInTheDocument();
  });

  it('should be able to change value', async () => {
    const onChange = jest.fn();

    const { container } = await waitForComponent(
      <TextareaField label="LABEL" value="" onChange={onChange} />,
      { wrapper: TestComponentWrapper },
    );
    expect(onChange).not.toHaveBeenCalled();

    const textarea = getByRole(container, 'textbox');
    expect(textarea).toHaveTextContent('');

    // type some text
    act(() => {
      fireEvent.change(textarea, { target: { value: 'TEXT' } });
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('TEXT');
  });

  it('should not change value when disabled', async () => {
    const onChange = jest.fn();

    const { container } = await waitForComponent(
      <TextareaField label="LABEL" value="" disabled onChange={onChange} />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();
    expect(onChange).not.toHaveBeenCalled();

    const textarea = getByRole(container, 'textbox');
    expect(textarea).toBeInTheDocument();
    expect(textarea).toHaveAttribute('disabled', '');

    // type some text
    act(() => {
      fireEvent.change(textarea, { target: { value: 'NEW_TEXT' } });
    });

    expect(onChange).not.toHaveBeenCalled();
  });
});
