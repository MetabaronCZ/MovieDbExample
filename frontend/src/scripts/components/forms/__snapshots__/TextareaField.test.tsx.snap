// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/forms/TextareaField should not change value when disabled 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 4px;
  align-items: flex-start;
  justify-content: start;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
  resize: none;
}

.c3::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c3:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c3:disabled {
  color: #98c3d9;
}

.c4 {
  width: 100%;
  padding: 0 12px;
  border: 2px solid #002d56;
  background: #fff;
  padding-top: 4px;
  padding-bottom: 4px;
  height: calc(80px + 2 * 4px);
}

.c4:focus {
  outline: 1px dashed #002d56;
}

.c4:disabled {
  border-color: #98c3d9;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="textarea-3"
      >
        LABEL
      </label>
      <textarea
        autocomplete="one-time-code"
        class="c3 c4"
        disabled=""
        id="textarea-3"
      />
    </div>
  </fieldset>
</div>
`;

exports[`components/forms/TextareaField should render error 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 4px;
  align-items: flex-start;
  justify-content: start;
}

.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  color: #f00;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
  color: #f00;
  resize: none;
}

.c3::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c3:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c3:disabled {
  color: #98c3d9;
}

.c4 {
  width: 100%;
  padding: 0 12px;
  border: 2px solid #002d56;
  background: #fff;
  border-color: #f00;
  padding-top: 4px;
  padding-bottom: 4px;
  height: calc(80px + 2 * 4px);
}

.c4:focus {
  outline: 1px dashed #002d56;
}

.c4:disabled {
  border-color: #98c3d9;
}

<div>
  <fieldset
    class="c0"
    data-invalid="true"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="textarea-0"
      >
        LABEL
      </label>
      <textarea
        autocomplete="one-time-code"
        class="c3 c4"
        id="textarea-0"
      >
        VALUE
      </textarea>
    </div>
    <p
      class="c5"
    >
      SOME-ERROR
    </p>
  </fieldset>
</div>
`;

exports[`components/forms/TextareaField should render info text 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 4px;
  align-items: flex-start;
  justify-content: start;
}

.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  color: #98c3d9;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
  resize: none;
}

.c3::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c3:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c3:disabled {
  color: #98c3d9;
}

.c4 {
  width: 100%;
  padding: 0 12px;
  border: 2px solid #002d56;
  background: #fff;
  padding-top: 4px;
  padding-bottom: 4px;
  height: calc(80px + 2 * 4px);
}

.c4:focus {
  outline: 1px dashed #002d56;
}

.c4:disabled {
  border-color: #98c3d9;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="textarea-1"
      >
        LABEL
      </label>
      <textarea
        autocomplete="one-time-code"
        class="c3 c4"
        id="textarea-1"
      >
        VALUE
      </textarea>
    </div>
    <p
      class="c5"
    >
      SOME-INFO
    </p>
  </fieldset>
</div>
`;

exports[`components/forms/TextareaField should render properly 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c3 {
  color: #f00;
}

.c3::after {
  content: '*';
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 4px;
  align-items: flex-start;
  justify-content: start;
}

.c4 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
  resize: none;
}

.c4::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c4:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c4:disabled {
  color: #98c3d9;
}

.c5 {
  width: 100%;
  padding: 0 12px;
  border: 2px solid #002d56;
  background: #fff;
  padding-top: 4px;
  padding-bottom: 4px;
  height: calc(8px + 2 * 4px);
}

.c5:focus {
  outline: 1px dashed #002d56;
}

.c5:disabled {
  border-color: #98c3d9;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="ID"
      >
        <span
          class="c3"
        />
        LABEL
      </label>
      <textarea
        autocomplete="one-time-code"
        class="c4 c5"
        id="ID"
        maxlength="100"
        name="NAME"
        placeholder="PLACEHOLDER"
      >
        VALUE
      </textarea>
    </div>
  </fieldset>
</div>
`;
